pgjdbc/src/main/java/org/postgresql/util/LazyCleaner.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+/* changes were made to move it into the org.postgresql.util package
+ *
+ * Copyright 2022 Juan Lopes
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.postgresql.util;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class LazyCleaner {
+  private static final Logger LOGGER = Logger.getLogger(LazyCleaner.class.getName());
+
+  public interface Cleanable {
+    void clean();
+  }
+
+  public interface CleaningAction {
+    void onClean(boolean leak) throws Exception;
```

Comment:
I wonder if adding generics would prevent `throws Exception` on the use site.

```suggestion
  public interface CleaningAction <T extends Throwable> {
    void onClean(boolean leak) throws T;
```

Reply:
seems to work, thanks for the suggestion